File: packages/swt3-ai/swt3_ai/exporters/otel.py
"""SWT3 AI Witness SDK -- OpenTelemetry Exporter.

Exports SWT3 witness anchors as OpenTelemetry spans, allowing them to
flow into existing observability pipelines (Datadog, Grafana, Jaeger,
Honeycomb, etc.).

Usage:
    from swt3_ai import Witness
    from swt3_ai.exporters.otel import OTelExporter

    exporter = OTelExporter(tracer_name="swt3-witness")
    witness = Witness(..., on_flush=exporter.export)

    # Anchors now appear as OTel spans alongside your existing traces.

Requires: pip install opentelemetry-api
    Or: pip install swt3-ai[otel]
"""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from ..types import WitnessPayload, WitnessReceipt

logger = logging.getLogger("swt3_ai.exporters.otel")

# GenAI semantic conventions: known provider prefixes -> gen_ai.system value
_GENAI_SYSTEM_MAP: Dict[str, str] = {
    "openai": "openai",
    "gpt": "openai",
    "anthropic": "anthropic",
    "claude": "anthropic",
    "bedrock": "aws.bedrock",
    "aws": "aws.bedrock",
    "azure": "az.ai.inference",
    "gemini": "google.gemini",
    "google": "google.gemini",
    "vertex": "google.vertex_ai",
    "ollama": "ollama",
    "vllm": "vllm",
    "litellm": "litellm",
    "mistral": "mistral_ai",
    "cohere": "cohere",
    "replicate": "replicate",
    "huggingface": "huggingface",
}


def _infer_genai_system(provider: Optional[str], model_id: Optional[str]) -> Optional[str]:
    """Infer gen_ai.system from provider or model_id. Returns None if unknown."""
    for source in (provider, model_id):
        if source:
            lower = source.lower()
            for prefix, system in _GENAI_SYSTEM_MAP.items():
                if lower.startswith(prefix):
                    return system
    return None


class OTelExporter:
    """Export SWT3 witness data as OpenTelemetry spans.

    Each flushed anchor becomes a span with swt3.* attributes. Pass
    ``exporter.export`` as the ``on_flush`` callback when creating
    a Witness instance.
    """

    def __init__(
        self,
        tracer_name: str = "swt3-ai",
        service_name: str = "swt3-witness",
    ) -> None:
        try:
            from opentelemetry import trace
            self._trace = trace
            self._tracer = trace.get_tracer(tracer_name)
        except ImportError:
            raise ImportError(
                "opentelemetry-api is required for OTel export. "
                "Install with: pip install opentelemetry-api  "
                "Or: pip install swt3-ai[otel]"
            )
        self._service_name = service_name

    def export(
        self,
        payloads: List[Any],
        receipts: List[Any],
    ) -> None:
        """Callback for Witness on_flush. Creates one span per anchor.

        This method is safe to pass directly as the on_flush callback:
            witness = Witness(..., on_flush=exporter.export)
        """
        for i, payload in enumerate(payloads):
            receipt = receipts[i] if i < len(receipts) else None
            attrs = self._span_attributes(payload, receipt)
            span_name = f"swt3.witness.{getattr(payload, 'procedure_id', 'unknown')}"

            with self._tracer.start_as_current_span(span_name, attributes=attrs) as span:
                if receipt and not getattr(receipt, "ok", True):
                    span.set_status(
                        self._trace.StatusCode.ERROR,
                        getattr(receipt, "error", "") or "",
                    )

    def _span_attributes(
        self,
        payload: Any,
        receipt: Optional[Any] = None,
    ) -> Dict[str, Any]:
        """Build OTel span attributes from a witness payload and receipt."""
        attrs: Dict[str, Any] = {}

        # Core anchor fields
        _set(attrs, "swt3.procedure_id", getattr(payload, "procedure_id", None))
        _set(attrs, "swt3.clearing_level", getattr(payload, "clearing_level", None))
        _set(attrs, "swt3.fingerprint", getattr(payload, "anchor_fingerprint", None))
        _set(attrs, "swt3.epoch", getattr(payload, "anchor_epoch", None))

        # Factors
        _set(attrs, "swt3.factor_a", getattr(payload, "factor_a", None))
        _set(attrs, "swt3.factor_b", getattr(payload, "factor_b", None))
        _set(attrs, "swt3.factor_c", getattr(payload, "factor_c", None))

        # AI metadata (may be cleared depending on level)
        _set(attrs, "swt3.model_id", getattr(payload, "ai_model_id", None))
        _set(attrs, "swt3.latency_ms", getattr(payload, "ai_latency_ms", None))

        # Identity (survives all clearing levels)
        _set(attrs, "swt3.agent_id", getattr(payload, "agent_id", None))
        _set(attrs, "swt3.cycle_id", getattr(payload, "cycle_id", None))

        # Receipt fields (from server response)
        if receipt:
            _set(attrs, "swt3.verdict", getattr(receipt, "verdict", None))
            _set(attrs, "swt3.anchor", getattr(receipt, "swt3_anchor", None))

        # GenAI semantic conventions (gen_ai.*)
        # Maps SWT3 witness data to OTel GenAI attributes for observability
        # pipeline compatibility (Datadog, Grafana, Honeycomb GenAI views).
        ctx = getattr(payload, "ai_context", None) or {}
        provider = ctx.get("provider") if isinstance(ctx, dict) else None
        model_id = getattr(payload, "ai_model_id", None)
        clearing = getattr(payload, "clearing_level", 1)

        system = _infer_genai_system(provider, model_id)
        if system:
            attrs["gen_ai.system"] = system

        if model_id:
            attrs["gen_ai.request.model"] = model_id

        # Token data only at CL0-1 (cleared at CL2+)
        if clearing is not None and clearing <= 1 and isinstance(ctx, dict):
            input_tokens = ctx.get("input_tokens") or ctx.get("tokens_in")
            output_tokens = ctx.get("output_tokens") or ctx.get("tokens_out")
            if input_tokens is not None:
                attrs["gen_ai.usage.input_tokens"] = input_tokens
            if output_tokens is not None:
                attrs["gen_ai.usage.output_tokens"] = output_tokens

        return attrs


def _set(attrs: Dict[str, Any], key: str, value: Any) -> None:
    """Set an attribute only if the value is not None."""
    if value is not None:
        attrs[key] = value

File: packages/swt3-ai-ts/src/exporters/otel.ts
/**
 * SWT3 AI Witness SDK -- OpenTelemetry Exporter.
 *
 * Exports SWT3 witness anchors as OpenTelemetry spans, allowing them to
 * flow into existing observability pipelines (Datadog, Grafana, Jaeger,
 * Honeycomb, etc.).
 *
 * Usage:
 *   import { Witness } from "@tenova/swt3-ai";
 *   import { OTelExporter } from "@tenova/swt3-ai/exporters/otel";
 *
 *   const exporter = new OTelExporter({ tracerName: "swt3-witness" });
 *   const witness = new Witness({ ..., onFlush: exporter.export.bind(exporter) });
 *
 * Requires: npm install @opentelemetry/api
 *
 * Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
 */

import type { WitnessPayload, WitnessReceipt } from "../types.js";

// GenAI semantic conventions: known provider prefixes -> gen_ai.system value
const GENAI_SYSTEM_MAP: Record<string, string> = {
  openai: "openai",
  gpt: "openai",
  anthropic: "anthropic",
  claude: "anthropic",
  bedrock: "aws.bedrock",
  aws: "aws.bedrock",
  azure: "az.ai.inference",
  gemini: "google.gemini",
  google: "google.gemini",
  vertex: "google.vertex_ai",
  ollama: "ollama",
  vllm: "vllm",
  litellm: "litellm",
  mistral: "mistral_ai",
  cohere: "cohere",
  replicate: "replicate",
  huggingface: "huggingface",
};

function inferGenaiSystem(provider?: string, modelId?: string): string | undefined {
  for (const source of [provider, modelId]) {
    if (source) {
      const lower = source.toLowerCase();
      for (const [prefix, system] of Object.entries(GENAI_SYSTEM_MAP)) {
        if (lower.startsWith(prefix)) return system;
      }
    }
  }
  return undefined;
}

export interface OTelExporterOptions {
  tracerName?: string;
  serviceName?: string;
}

export class OTelExporter {
  private tracer: any;
  private traceModule: any;

  constructor(options?: OTelExporterOptions) {
    const tracerName = options?.tracerName ?? "swt3-ai";
    try {
      // Dynamic require to avoid hard dependency
      const trace = require("@opentelemetry/api").trace;
      this.traceModule = require("@opentelemetry/api");
      this.tracer = trace.getTracer(tracerName);
    } catch {
      throw new Error(
        "@opentelemetry/api is required for OTel export. " +
        "Install with: npm install @opentelemetry/api"
      );
    }
  }

  /**
   * Callback for Witness onFlush. Creates one span per anchor.
   *
   * Pass as: onFlush: exporter.export.bind(exporter)
   */
  export(payloads: WitnessPayload[], receipts: WitnessReceipt[]): void {
    for (let i = 0; i < payloads.length; i++) {
      const payload = payloads[i];
      const receipt = receipts[i] ?? null;
      const spanName = `swt3.witness.${payload.procedure_id ?? "unknown"}`;
      const attrs = this.spanAttributes(payload, receipt);

      const span = this.tracer.startSpan(spanName, { attributes: attrs });

      if (receipt && !(receipt as any).ok) {
        span.setStatus({
          code: this.traceModule.SpanStatusCode.ERROR,
          message: (receipt as any).error ?? "",
        });
      }

      span.end();
    }
  }

  private spanAttributes(
    payload: WitnessPayload,
    receipt: WitnessReceipt | null,
  ): Record<string, string | number | boolean> {
    const attrs: Record<string, string | number | boolean> = {};

    // Core anchor fields
    if (payload.procedure_id) attrs["swt3.procedure_id"] = payload.procedure_id;
    if (payload.clearing_level != null) attrs["swt3.clearing_level"] = payload.clearing_level;
    if (payload.anchor_fingerprint) attrs["swt3.fingerprint"] = payload.anchor_fingerprint;
    if (payload.anchor_epoch) attrs["swt3.epoch"] = payload.anchor_epoch;

    // Factors
    if (payload.factor_a != null) attrs["swt3.factor_a"] = payload.factor_a;
    if (payload.factor_b != null) attrs["swt3.factor_b"] = payload.factor_b;
    if (payload.factor_c != null) attrs["swt3.factor_c"] = payload.factor_c;

    // AI metadata (may be cleared depending on level)
    if (payload.ai_model_id) attrs["swt3.model_id"] = payload.ai_model_id;
    if (payload.ai_latency_ms != null) attrs["swt3.latency_ms"] = payload.ai_latency_ms;

    // Identity (survives all clearing levels)
    if (payload.agent_id) attrs["swt3.agent_id"] = payload.agent_id;
    if (payload.cycle_id) attrs["swt3.cycle_id"] = payload.cycle_id;

    // Receipt fields
    if (receipt) {
      if (receipt.verdict) attrs["swt3.verdict"] = receipt.verdict;
      if (receipt.swt3_anchor) attrs["swt3.anchor"] = receipt.swt3_anchor;
    }

    // GenAI semantic conventions (gen_ai.*)
    // Maps SWT3 witness data to OTel GenAI attributes for observability
    // pipeline compatibility (Datadog, Grafana, Honeycomb GenAI views).
    const ctx = payload.ai_context as Record<string, any> | undefined;
    const provider = ctx?.provider as string | undefined;
    const modelId = payload.ai_model_id;
    const clearing = payload.clearing_level;

    const system = inferGenaiSystem(provider, modelId);
    if (system) attrs["gen_ai.system"] = system;

    if (modelId) attrs["gen_ai.request.model"] = modelId;

    // Token data only at CL0-1 (cleared at CL2+)
    if (clearing != null && clearing <= 1 && ctx) {
      const inputTokens = ctx.input_tokens ?? ctx.tokens_in;
      const outputTokens = ctx.output_tokens ?? ctx.tokens_out;
      if (inputTokens != null) attrs["gen_ai.usage.input_tokens"] = inputTokens;
      if (outputTokens != null) attrs["gen_ai.usage.output_tokens"] = outputTokens;
    }

    return attrs;
  }
}
